CRM/Price/BAO/PriceSet.php

```diff
@@ -1747,6 +1767,205 @@ public static function getNonDeductibleAmountFromPriceSet($priceSetId, $lineItem
+   *
+   * @param int $membership_id The membership that is being renewed.
+   *
+   * @return int
+   *     e.g., 9. To indicate that price set #9 is the last price set that as
+   *           used on contact #3.
+   */
+  public static function getLastPriceSetUsed($membership_id) {
+    $sql = "
+SELECT p.contribution_id, c.receive_Date, pf.price_set_id
+  FROM civicrm_membership m
+    LEFT JOIN civicrm_membership_payment p
+           ON p.membership_id = m.id
+     LEFT JOIN civicrm_contribution c
+            ON c.id = p.contribution_id
+    LEFT JOIN civicrm_line_item li
+           ON li.contribution_id = p.contribution_id
+          AND li.entity_table = 'civicrm_membership'
+          AND li.entity_id = m.id
+    LEFT JOIN civicrm_price_field_value pfv
+           ON pfv.id = li.price_field_value_id
+          AND pfv.is_active = 1
+    LEFT JOIN civicrm_price_field pf
+           ON pf.id = pfv.price_field_id
+          AND pf.is_active = 1
+    LEFT JOIN civicrm_price_set ps
+           ON ps.id = pf.price_set_id
+          AND ps.is_active = 1
+ WHERE m.id = 34
+       ORDER BY c.receive_date DESC, p.contribution_id DESC
+    ";
+
+    $params = array(1 => array($membership_id, 'Integer'));
+    $dao = self::executeQuery($sql, $params);
+    $activateByDefault = TRUE;
+    while ($dao->fetch()) {
+      // last price set used on this membership (could be multiple over the course of time)
+      // only select "price set view", if last contribution was for the price set.
+      if ($dao->price_set_id) {
+        return array(
+          'price_set_id' => $dao->price_set_id,
+          'price_set_is_through_contribution' => $activateByDefault,
+        );
+      }
+      $activateByDefault = FALSE;
+    }
+
+    // nothing found.  we want to pick, by default the price set that will cover
+    // the most of this contact's members of organizations.
+    $sql = "
+SELECT ps.id, group_concat(DISTINCT member_of_contact_id) AS member_of_contact_ids
+ FROM civicrm_price_set ps
+   INNER JOIN civicrm_price_field pf ON pf.price_set_id = ps.id AND pf.is_active = 1
+   INNER JOIN civicrm_price_field_value pfv ON pfv.price_field_id = pf.id AND pfv.is_active = 1
+   INNER JOIN civicrm_membership_type mt ON mt.id = pfv.membership_type_id
+ WHERE ps.is_active = 1
+ GROUP BY ps.id
+ ";
+
+    $dao = self::executeQuery($sql);
+    $orgsByPriceSet = array();
+    while ($dao->fetch()) {
+      $orgsByPriceSet[$dao->id] = explode(",", $dao->member_of_contact_ids);
+    }
+
+    $memberhsip_type_id = CRM_Member_BAO_Membership::getFieldValue('CRM_Member_BAO_Membership', $membership_id, 'membership_type_id');
+    $member_of_contact_id = CRM_Member_BAO_MembershipType::getFieldValue('CRM_Member_BAO_MembershipType', $memberhsip_type_id, 'member_of_contact_id');
+    $contact_id = CRM_Member_BAO_Membership::getFieldValue('CRM_Member_BAO_Membership', $membership_id, 'contact_id');
+
+    // we've got an array, above, of all orgs mapped on price set id.
```

Comment:
The client I'm working with has several price sets to pick from, each price set covering several orgs & types, and with overlap in various membership orgs & types between the price sets.  Typically, one price set handles all of a member's memberships.  Without going for the "most", there is an extra owness (sp?) on the user to "figure out" which price set would "work" the best.  It'd be workable without it, for sure, but not fully optimized from a user (or at least this client's) perspective.

It could alternatively be handled with a hook, but I figure 1) it would be more work through a hook, 2) although somewhat cumbersome SQL, going with the idea that if you switch to "price set", you likely do want to cover as many of the memberships as possible -- generally speaking, and 3) if it's good for this client, it would be likely be of value to others.